Detectors/FOCAL/simulation/src/Detector.cxx

```diff
@@ -165,6 +184,54 @@ Bool_t Detector::ProcessHits(FairVolume* v)
     }
   }
 
+  // Processing ECAL hits
+  if (TVirtualMC::GetMC()->CurrentMedium() == mMedSensECal) {
+    LOG(debug) << "We are in sensitive volume " << v->GetName() << ": " << TVirtualMC::GetMC()->CurrentVolPath();
+
+    double eloss = TVirtualMC::GetMC()->Edep() * 1e9; // energy in eV  (GeV->eV)
+    if (eloss < DBL_EPSILON) {
+      return false; // only process hits which actually deposit some energy in the FOCAL
+    }
+
+    // In case of new parent track create new track reference
+    auto o2stack = static_cast<o2::data::Stack*>(TVirtualMC::GetMC()->GetStack());
+    if (!mCurrentSuperparent->mHasTrackReference) {
+      float x, y, z, px, py, pz, e;
+      TVirtualMC::GetMC()->TrackPosition(x, y, z);
+      TVirtualMC::GetMC()->TrackMomentum(px, py, pz, e);
+      o2::TrackReference trackref(x, y, z, px, py, pz, TVirtualMC::GetMC()->TrackLength(), TVirtualMC::GetMC()->TrackTime(), mCurrentParentID, GetDetId());
+      o2stack->addTrackReference(trackref);
+      mCurrentSuperparent->mHasTrackReference = true;
+    }
+
+    float posX, posY, posZ;
+    TVirtualMC::GetMC()->TrackPosition(posX, posY, posZ);
+
+    auto [indetector, col, row, layer, segment] = mGeometry->getVirtualInfo(posX, posY, posZ);
+
+    if (!indetector) {
+      // particle outside the detector
+      return true;
+    }
+
+    auto currenthit = FindHit(mCurrentParentID, col, row, layer);
```

Comment:
Please cross check that HCAL and ECAL layers are treated as separate layers in FindHit

Reply:
I belive that the method getVirtualInfo(...) from Geometry class is doing the job properly.